include/codec.h

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API header file for Audio Codec
+ *
+ * This file contains the Audio Codec APIs
+ */
+
+#ifndef __CODEC_H__
+#define __CODEC_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * PCM audio sample rates
+ */
+typedef enum {
+	CODEC_PCM_RATE_8K	= 8000,
+	CODEC_PCM_RATE_16K	= 16000,
+	CODEC_PCM_RATE_24K	= 24000,
+	CODEC_PCM_RATE_32K	= 32000,
+	CODEC_PCM_RATE_44P1K	= 44100,
+	CODEC_PCM_RATE_48K	= 48000,
+	CODEC_PCM_RATE_96K	= 96000,
+	CODEC_PCM_RATE_192K	= 192000,
+} codec_pcm_rate_t;
+
+/**
+ * PCM audio sample bit widths
+ */
+typedef enum {
+	CODEC_PCM_WIDTH_16_BITS	= 16,
+	CODEC_PCM_WIDTH_20_BITS	= 20,
+	CODEC_PCM_WIDTH_24_BITS	= 24,
+	CODEC_PCM_WIDTH_32_BITS	= 32,
+} codec_pcm_width_t;
+
+/**
+ * Digital Audio Interface (DAI) type
+ */
+typedef enum {
+	CODEC_DAI_TYPE_I2S,
+	CODEC_DAI_TYPE_INVALID,
+} codec_dai_type_t;
+
+/**
+ * Digital Audio Interface (DAI) format
+ */
+typedef enum {
+	CODEC_DAI_MODE_I2S,	/* standard I2S mode */
+	CODEC_DAI_MODE_DSP,	/* DSP mode */
+	CODEC_DAI_MODE_RJF,	/* Right justified mode */
+	CODEC_DAI_MODE_LJF,	/* Left justified mode */
+	CODEC_DAI_MODE_TDM,	/* Multi-channel TDM mode */
+	CODEC_DAI_MODE_INVALID,
+} codec_dai_mode_t;
+
+/**
+ * Codec configuration parameters
+ */
+struct codec_cfg {
+	u32_t			mclk_freq;	/* MCLK input frequency in Hz */
+	codec_pcm_rate_t	pcm_rate;	/* PCM sample rate in Hz */
+	codec_pcm_width_t	pcm_width;	/* PCM sample size in bits */
+	codec_dai_type_t	dai_type;	/* Digital interface type */
+	codec_dai_mode_t	dai_mode;	/* Digital interface mode */
+	u32_t			channels;	/* Number of audio channels */
+	bool 			dai_master;	/* Is codec the DAI master? */
+};
```

Comment:
Btw, is this suppose to be PCM only interface? Also what is the expected format? I suspect we want something that allows any audio codec to be registered by a driver and then have the audio subsystem expose this in a generic way so the applications don't have to interface directly with drivers.  We may actually have the audio API to build pipelines where we can chain the data which would allow not only encoding and decoding but also allow to control input, output, mux, demux, resampling, transcoding, etc. Perhaps we should first create an issue and then we start adding requirements which should make clear what sort of APIs the audio will need.

Reply:
Currently it is only PCM interface -- the hardware I'm working with supports only PCM interface. If there are others running Zephyr on hardware supporting other interfaces, we could extend the API to add support for those as I responded earlier.
There is no specific expected format, the API is intended to allow specifying the format (pcm_rate, pcm_width, etc.). 
Regarding audio pipelines, I think they should not be in Zephyr. Instead, we should leverage the other Linux Foundation project https://www.sofproject.org/. So Zephyr for hardware drivers and RTOS services, SOF for audio processing.

Reply:
If this is supposed to be a generic audio codec interface, I'd suggest to not make it I2S specific at all. Rather it should only contain what is common for all expected audio codecs. One could think in OOP terms, where this would act as a base and an then there would be a specific I2S specialization.
That would separate codecs from pipeline and routing.

Reply:
@Vudentz You are welcome to define those APIs as appropriate. For now, I am focused on enabling a codec driver and that's the API I'm trying to define.
@eisbaw Yes this is supposed to be generic audio codec interface. I am in favor of separate the codec configuration and the interface configuration (I2S). Please provide any specific suggestion you have on a structure definition. What is OOP by the way?

Reply:
Instead of re-defining digital interface configuration, I made a change to reuse existing definitions.
I have included `i2s.h` as this is the only digital interface that seems to exist in the code base. When other types of interfaces are available, their header can be included and the configuration structure can be included in `a_codec_cfg_t` union.
I hope this is a good compromise that is acceptable to everyone.